src/proto/streams/streams.rs

```diff
@@ -402,44 +404,61 @@ where
 
-        let res = {
-            let stream = match me.store.find_mut(&id) {
-                Some(stream) => stream.key(),
-                None => return Err(RecvError::Connection(Reason::PROTOCOL_ERROR)),
-            };
+        // First, ensure that the initiating stream is still in a valid state.
+        let parent_key = match me.store.find_mut(&id) {
+            Some(stream) => {
+                if id > me.actions.recv.max_stream_id() {
+                    trace!("id ({:?}) > max_stream_id ({:?}), ignoring PUSH_PROMISE", id, me.actions.recv.max_stream_id());
+                    return Ok(());
+                }
 
-            if me.counts.peer().is_server() {
-                // The remote is a client and cannot reserve
-                trace!("recv_push_promise; error remote is client");
-                return Err(RecvError::Connection(Reason::PROTOCOL_ERROR));
+                // The stream must be receive open
+                stream.state.ensure_recv_open()?;
+                stream.key()
             }
-
-            me.actions.recv.recv_push_promise(frame,
-                                              &me.actions.send,
-                                              stream,
-                                              &mut me.store)
+            None => return Err(RecvError::Connection(Reason::PROTOCOL_ERROR)),
         };
 
-        if let Err(err) = res {
-            if let Some(ref mut new_stream) = me.store.find_mut(&promised_id) {
+        // Ensure that we can reserve streams
+        me.actions.recv.ensure_can_reserve()?;
 
-                let mut send_buffer = self.send_buffer.inner.lock().unwrap();
-                me.actions.reset_on_recv_stream_err(
-                    &mut *send_buffer,
-                    new_stream,
-                    &mut me.counts,
-                    Err(err))
-            } else {
-                // If there was a stream error, the stream should have been stored
-                // so we can track sending a reset.
-                //
-                // Otherwise, this MUST be an connection error.
-                assert!(!err.is_stream_error());
-                Err(err)
-            }
-        } else {
-            res
+        // Next, open the stream.
+        if me.actions.recv.open(promised_id, true, &mut me.counts)?.is_none() {
+            return Ok(());
         }
+
+        // Create a scope
+        let child_key = {
+            // Create state for the stream
+            let stream = me.store.insert(promised_id, {
+                Stream::new(
+                    promised_id,
+                    me.actions.send.init_window_sz(),
+                    me.actions.recv.init_window_sz())
+            });
+
+            let actions = &mut me.actions;
+
+            me.counts.transition(stream, |counts, stream| {
+                let res = actions.recv.recv_push_promise(frame, stream);
+
+                let mut send_buffer = self.send_buffer.inner.lock().unwrap();
```

Comment:
We don't need to take this lock if there wasn't an error. The previous code checked before locking.

Reply:
I don't disagree with you, but I opted to unconditionally take the lock because:

* There are plenty of other places in the code that unconditionally take the lock, and any fixes should be applied globally.
* The function being called takes the `res` and conditionally does work on error, it seems like it would be better to pass the mutex in vs. the buffer. Either way, it felt weird conditionally calling a fn based on the logic that the fn is supposed to deal with for you.
* There are no measurements that unconditionally taking the lock there matters and it is an edge case.

Given this, I opted for simpler code in favor of fixing it universally at a later time. If you disagree, I can revert back.